### PR TITLE
Avoid duplicate paths if path is not a directory

### PR DESCRIPTION
**Pull request by Sophist-UK**
Form path arrays contain duplicates when they shouldn't.

Current code:
```
foreach ($new as $path)
{
	if (!in_array($path, $paths))
	{
		array_unshift($paths, trim($path));
	}

	if (!is_dir($path))
	{
		array_unshift($paths, trim($path));
	}
}
```

Proposed code:
```
foreach ($new as $path)
{
	$path = trim($path);
	if (!in_array($path, $paths) && is_dir($path))
	{
		array_unshift($paths, $path);
	}
}
```

### Summary of Changes

Path arrays currently have duplicate path in them - this PR removes the code which creates duplicates.

### Testing Instructions
```
$form = new JForm('test');
var_dump($form->addFieldPath('/')); // Exists
var_dump($form->addFieldPath('/directory_doesnt_exist')); // Does not exist
```
### Existing result

Returned array shows `/` and  `/directory_doesnt_exist` twice.

### Fixed result

Returned array shows `/` once and  `/directory_doesnt_exist` is omitted.

### Documentation Changes Required

None.

### CODE REVIEW

File: libraries/src/Form/FormHelper.php [hunk 1]
```diff
@@ -331,14 +331,10 @@ protected static function addPath($entity, $new = null)
 		// Add the new paths to the stack if not already there.
 		foreach ($new as $path)
 		{
-			if (!in_array($path, $paths))
+			$path = trim($path);
+			if (!in_array($path, $paths) && is_dir($path))
```
Comment (Quy) [inline libraries/src/Form/FormHelper.php]: Please add blank line before `if` statement.

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on b57ddc39640a975484d673e42d0a6206b8776ffc<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19781">issues.joomla.org/tracker/joomla-cms/19781</a>.</sub>

**Comment by anuragteapot** [comment]
I have tested this item :white_check_mark: successfully on b57ddc39640a975484d673e42d0a6206b8776ffc<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19781">issues.joomla.org/tracker/joomla-cms/19781</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19781">issues.joomla.org/tracker/joomla-cms/19781</a>.</sub>

**Comment by mbabker** [comment]
The unit test failure needs to be addressed before this can be merged.

**Comment by Sophist-UK** [comment]
Apologies - I had not spotted that. Working on it now.

File: tests/unit/suites/libraries/joomla/form/JFormTest.php [hunk 2]
```diff
@@ -2201,6 +2201,9 @@ public function testSyncPaths()
 		$fieldPaths = JForm::addFieldPath();
 		$formPaths = JForm::addFormPath();
 		$rulePaths = JForm::addRulePath();
+var_dump('fieldpaths',$fieldPaths);
```
Comment (Bakual) [inline tests/unit/suites/libraries/joomla/form/JFormTest.php]: This obviously is wrong 😄

Comment (Sophist-UK) [reply tests/unit/suites/libraries/joomla/form/JFormTest.php]: No - it's correct for what I wanted to do, which is to see what is in these variables. I have no way of running these tests locally without a LOT of setup.

But obviously it is not code to be included in the solution.

However, this is what is in these variables:

**fieldpaths**
/home/travis/build/joomla/joomla-cms/tests/unit/suites/libraries/joomla/form
/home/travis/build/joomla/joomla-cms/tests/unit/suites/libraries/joomla/form/_testfields
/home/travis/build/joomla/joomla-cms/libraries/joomla/form/fields
/home/travis/build/joomla/joomla-cms/libraries/src/Form/fields

**formpaths**
/home/travis/build/joomla/joomla-cms/tests/unit/suites/libraries/joomla/form
/home/travis/build/joomla/joomla-cms/libraries/src/Form/forms

**rulepaths**
/home/travis/build/joomla/joomla-cms/tests/unit/suites/libraries/joomla/form/_testrules
/home/travis/build/joomla/joomla-cms/tests/unit/suites/libraries/joomla/form
/home/travis/build/joomla/joomla-cms/libraries/src/Form/rules

I have no idea why these strings are in there, nor why field1/2/3 etc. should be in there, nor why my change means they are no longer in there.

I need some help to understand this, and then I can fix the tests.

(The only tests I did was to discover that the original code resulted in weird arrays, fix the code and see that the arrays on my real Joomla system were now as expected.)

**Comment by Sophist-UK** [comment]
The tests I did when coding was to discover that the original code resulted in weird arrays on my Joomla system, fix the code and see that the arrays on my real Joomla system were now as expected.

This is what is in these variables:

**fieldpaths**
/home/travis/build/joomla/joomla-cms/tests/unit/suites/libraries/joomla/form
/home/travis/build/joomla/joomla-cms/tests/unit/suites/libraries/joomla/form/_testfields
/home/travis/build/joomla/joomla-cms/libraries/joomla/form/fields
/home/travis/build/joomla/joomla-cms/libraries/src/Form/fields

**formpaths**
/home/travis/build/joomla/joomla-cms/tests/unit/suites/libraries/joomla/form
/home/travis/build/joomla/joomla-cms/libraries/src/Form/forms

**rulepaths**
/home/travis/build/joomla/joomla-cms/tests/unit/suites/libraries/joomla/form/_testrules
/home/travis/build/joomla/joomla-cms/tests/unit/suites/libraries/joomla/form
/home/travis/build/joomla/joomla-cms/libraries/src/Form/rules

I have no idea why these strings are in there, nor why field1/2/3 etc. *should* be in there, nor why my change means they are no longer in there.

I need some help to understand this, and then I can fix the tests.

**Comment by Sophist-UK** [comment]
Ok - I found where the paths came from - XML in JFormDataHelper.php.

These are not being added because the directory for e.g. field1 doesn't exist.

It should be easy to make the existing tests work by removing the check that the path is to a directory and allow adding files as well as directories, however I am unclear whether these arrays should be of directories only, or whether we should also include files. So I need a steer on this.

P.S. The tests only test that the paths are added from XML - they do not currently check the correct operation of the addXxxxxPath i.e. that directories are added and files and non-existent paths are not. I will add these tests when I fix this.

**Comment by Sophist-UK** [comment]
This should now be mergable.

**Review by mbabker — APPROVED** [review]

**Comment by viocassel** [comment]
I have tested this item :white_check_mark: successfully on b78f8798b8693860323677c6f66149adf0526b86<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19781">issues.joomla.org/tracker/joomla-cms/19781</a>.</sub>

**Comment by ghost** [comment]
@Quy can you please retest?

**Comment by SharkyKZ** [comment]
I have tested this item :white_check_mark: successfully on b78f8798b8693860323677c6f66149adf0526b86<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19781">issues.joomla.org/tracker/joomla-cms/19781</a>.</sub>

**Comment by SharkyKZ** [comment]
RTC.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19781">issues.joomla.org/tracker/joomla-cms/19781</a>.</sub>

**Comment by SharkyKZ** [comment]
Though I don't understand why do we add non-existing paths anyways.

**Comment by mbabker** [comment]
> Though I don't understand why do we add non-existing paths anyways.

It doesn't hurt anything to have non-existing paths in the registry as long as at runtime when that registry is used the paths are checked to make sure they exist.  Otherwise you're doing a check when adding to the registry and when the registry is used, and that comes with a performance hit because it involves filesystem I/O.

**Comment by HLeithner** [comment]
Finally...

Thank you for your optimization.

**Comment by Sophist-UK** [comment]
Wow - submitted on 25 Feb 2018, and merged on 11 December 2019 _**only**_ 23 months later. Is this a record?

P.S. It was only a small fix, but I am glad to be able to contribute.

**Pull request merged**